tsl/src/compression/compression.c

```diff
@@ -1836,6 +1836,62 @@ tsl_compressed_data_out(PG_FUNCTION_ARGS)
 	PG_RETURN_CSTRING(encoded);
 }
 
+Oid
+get_index_on_compressed_chunk(Chunk *uncompressed_chunk, Chunk *compressed_chunk)
+{
+	int32 srcht_id = uncompressed_chunk->fd.hypertable_id;
+	List *htcols_list = ts_hypertable_compression_get(srcht_id);
+	ListCell *lc;
+	int htcols_listlen = list_length(htcols_list);
+
+	const ColumnCompressionInfo **colinfo_array;
+	colinfo_array = palloc(sizeof(ColumnCompressionInfo *) * htcols_listlen);
+
+	int i = 0;
+	foreach (lc, htcols_list)
+	{
+		FormData_hypertable_compression *fd = (FormData_hypertable_compression *) lfirst(lc);
+		colinfo_array[i++] = fd;
+	}
+
+	const ColumnCompressionInfo **keys;
+	int n_keys;
+	int16 *in_column_offsets = compress_chunk_populate_keys(uncompressed_chunk->table_id,
+															colinfo_array,
+															htcols_listlen,
+															&n_keys,
+															&keys);
+
+	Relation uncompressed_chunk_rel = table_open(uncompressed_chunk->table_id, ExclusiveLock);
+	Relation compressed_chunk_rel = table_open(compressed_chunk->table_id, ExclusiveLock);
+	TupleDesc compressed_rel_tupdesc = RelationGetDescr(compressed_chunk_rel);
+	TupleDesc uncompressed_rel_tupdesc = RelationGetDescr(uncompressed_chunk_rel);
+
+	RowCompressor row_compressor;
+	row_compressor_init(&row_compressor,
+						uncompressed_rel_tupdesc,
+						compressed_chunk_rel,
+						htcols_listlen,
+						colinfo_array,
+						in_column_offsets,
+						compressed_rel_tupdesc->natts,
+						true /*need_bistate*/,
+						true /*reset_sequence*/);
+
+	/*
+	 * Keep the ExclusiveLock on the compressed chunk. This lock will be requested
+	 * by recompression later on, both in the case of segmentwise recompression, and
+	 * in the case of decompress-compress. This implicitly locks the index too, so
+	 * it cannot be dropped in another session, which is what we want to prevent by
+	 * locking the compressed chunk here
+	 */
+	table_close(compressed_chunk_rel, NoLock);
+	table_close(uncompressed_chunk_rel, NoLock);
```

Comment:
Is it intended to keep the lock of the `uncompressed_chunk_rel` too? If so, could you add a comment why this is required?

Reply:
There is a comment just above.

Reply:
This comment only covers the `compressed_chunk_rel`. If the same applies to the `uncompressed_chunk_rel` it should be added to the comment.